src/ATen/native/xpu/sycl/LinearInt4.cpp

```diff
@@ -0,0 +1,167 @@
+#include <ATen/native/xpu/sycl/LinearInt4.h>
+#include <comm/SYCLContext.h>
+
+namespace at::native::xpu {
+
+template <typename scalar_t = at::Half, int block_size = 16>
+struct LinearInt4KernelFunctor : public __SYCL_KER_CONFIG_CONVENTION__ {
+  LinearInt4KernelFunctor(
+      scalar_t* A,
+      uint32_t* B,
+      scalar_t* C,
+      scalar_t* B_scale,
+      scalar_t* B_zero_point,
+      int m,
+      int n,
+      int k,
+      int lda,
+      int ldb,
+      int ldc)
+      : A(A),
+        B(B),
+        C(C),
+        B_scale(B_scale),
+        B_zero_point(B_zero_point),
+        m(m),
+        n(n),
+        k(k),
+        lda(lda),
+        ldb(ldb),
+        ldc(ldc) {}
+  void sycl_ker_config_convention(sycl::handler& cgh) {
+    // local_scan_ = sycl_local_acc_t<T>(N_, cgh);
+  }
+
+  void operator()(sycl::nd_item<1> item) const {
+    int constexpr Unroll = 2;
+    int constexpr SgSize = 16;
+    int constexpr TileK = 32;
+    int constexpr GroupK = SgSize * TileK;
+    int constexpr blocksize = 16;
+
+    int g_idx = item.get_group(0);
+    auto sg = item.get_sub_group();
+    int sg_id = sg.get_local_id()[0];
+    int g_n = g_idx;
+    auto sptr = B_scale + g_n * ldb;
+    auto bptr = B + g_n * k / 2;
+    auto aptr = A;
+    auto cptr = C + g_n;
+    if constexpr (std::is_same_v<scalar_t, sycl::half>) {
+      sycl::half2 tmpAcc = {0.f, 0.f};
+      for (int i = 0; i < k; i += GroupK * Unroll) {
+#pragma unroll
+        for (int iu = 0; iu < Unroll; iu++) {
+          uint8_t tmps8[TileK / 2];
+          *(sycl::vec<uint8_t, TileK / 2>*)tmps8 =
+              *(sycl::vec<uint8_t, TileK / 2>*)(bptr + sg_id * TileK / 2);
+          scalar_t scale = *(sptr + sg_id * TileK / blocksize);
+#pragma unroll
+          for (int ikk = 0; ikk < TileK; ikk += 2) {
+            sycl::half2 tmpA = *(sycl::half2*)&aptr[sg_id * TileK + ikk];
+            sycl::half2 tmpB = {
+                static_cast<int8_t>((tmps8[ikk / 2] & 0x0f) - 8),
+                static_cast<int8_t>((tmps8[ikk / 2] >> 4) - 8)};
+            tmpAcc += tmpA * tmpB * scale;
```

Comment:
is it possible to do vectorized load and shift with sycl? i don't know.
if not, i guess this is best perf that we can get so far. this line should be the major bottlenecks.

Reply:
depend on the IGC auto vectorization